types/three/examples/jsm/nodes/shadernode/ShaderNode.d.ts

```diff
@@ -163,6 +163,10 @@ export function nodeImmutable<T>(
     ...params: ProxiedTuple<GetConstructors<T>>
 ): ShaderNodeObject<ConstructedNode<T>>;
 
+export function tslFn<T, R extends Node = ShaderNodeObject<Node>>(
+    jsFunc: (args: T) => R,
+): T extends AnyObject ? (args: T) => R : () => R;
```

Comment:
Can you explain the return type? I see that `tslFn` does fancy things with the first argument, but not seeing a clear explanation for why we would just remove the first argument if it's not an object. I would be inclined to just return `(args: T) => R` unless I'm missing something.

Reply:
Thanks, that's helpful. Can you add tests to `test/unit/examples/jsm/nodes/ShaderNode/ShaderNode.ts` for both the single argument case and the no argument case? I feel like there might be a better way, but I want to see this way functioning first before suggesting alternatives.

Reply:
One idea that crossed my mind is using function overloads instead. Something like:

```ts
export function tslFn<R extends Node = ShaderNodeObject<Node>>(
    jsFunc: () => R,
): () => R;
export function tslFn<T, R extends Node = ShaderNodeObject<Node>>(
    jsFunc: (args: T) => R,
): (args: T) => R;
```

Another idea might be to make the function itself the generic, that might make it harder to manipulate the argument if needed.

Reply:
overloading looks great. I'll go with it. Thank you for the suggestion!

Reply:
added a test as well